plugins/statistics.koplugin/main.lua

```diff
@@ -544,26 +578,22 @@ function ReaderStatistics:getIdBookDB()
 end
 
 function ReaderStatistics:insertDB(id_book)
-    self.pages_stats[TimeVal:now().sec] = self.curr_page
-    if id_book == nil or util.tableSize(self.pages_stats) < 2 then
+    if id_book == nil or self.mem_read_pages < 2 then
         return
     end
-    local diff_time
+    local now_ts = TimeVal:now().sec
     local conn = SQ3.open(db_location)
-    local sorted_performance = {}
-    for time, pages in pairs(self.pages_stats) do
-        table.insert(sorted_performance, time)
-    end
-    table.sort(sorted_performance)
     conn:exec('BEGIN')
     local stmt = conn:prepare("INSERT OR IGNORE INTO page_stat VALUES(?, ?, ?, ?)")
-    for i=1, #sorted_performance - 1 do
-        diff_time = sorted_performance[i+1] - sorted_performance[i]
-        if diff_time >= self.page_min_read_sec then
-            stmt:reset():bind(id_book,
-                self.pages_stats[sorted_performance[i]],
-                sorted_performance[i],
-                math.min(diff_time, self.page_max_read_sec)):step()
+    for page, data_list in pairs(self.page_stat) do
+        for _, data_tuple in ipairs(data_list) do
+            -- See self.page_stat declaration above about the tuple's layout
```

Comment:
I don't know if previously, the entries were inserted (and so, shown when doing a SELECT) in timestamp order - and it shouldn't matter much with SQL (as we should use `order by timestamp` if needed) - but having them inserted by event order would be a small plus (when debugging, I remember I once did look at that).

Reply:
It was indeed inserted in *timestamp* order before, because of the mad temporary sorted array.

It's now supposedly completely random (`pairs` \o/), although it *tends* to be sorted in *page* order (pairs appears to be friendly to purely numerical hashes).

Reply:
As a small example:

```
INSERT INTO page_stat VALUES(1,8,1602022935,16);
INSERT INTO page_stat VALUES(1,9,1602022925,10);
INSERT INTO page_stat VALUES(1,9,1602022951,9);
INSERT INTO page_stat VALUES(1,8,1602022964,120);
INSERT INTO page_stat VALUES(1,13,1602024202,5);
INSERT INTO page_stat VALUES(1,14,1602024091,68);
INSERT INTO page_stat VALUES(1,15,1602024190,8);
INSERT INTO page_stat VALUES(1,16,1602024185,5);
INSERT INTO page_stat VALUES(1,17,1602024164,14);
INSERT INTO page_stat VALUES(1,17,1602024180,5);
INSERT INTO page_stat VALUES(1,9,1602024279,9);
INSERT INTO page_stat VALUES(1,10,1602024268,11);
INSERT INTO page_stat VALUES(1,10,1602024288,8);
INSERT INTO page_stat VALUES(1,11,1602024256,12);
INSERT INTO page_stat VALUES(1,11,1602024296,10);
INSERT INTO page_stat VALUES(1,12,1602024248,8);
INSERT INTO page_stat VALUES(1,12,1602024306,14);
INSERT INTO page_stat VALUES(1,13,1602024320,8);
```